wdl-analysis/src/queue.rs

```diff
@@ -723,6 +774,106 @@ where
 
         (index, document)
     }
+
+    /// Finds the definition location for an identifier at the given position.
+    ///
+    /// Searches the document and its imports (doesn't work now) for the
+    /// definition of the identifier at the specified position, returning
+    /// the location if found.
+    fn goto_definition(
+        &self,
+        document: Url,
+        position: SourcePosition,
+        encoding: SourcePositionEncoding,
+    ) -> Result<Option<GotoDefinitionLocation>> {
+        let graph = self.graph.read();
+
+        let index = graph
+            .get_index(&document)
+            .ok_or_else(|| anyhow!("document not found in graph"))?;
+
+        let lines = match graph.get(index).parse_state() {
```

Comment:
You should consolidate the `graph.get` calls here and later on. You only need to get the document once and error if it's not found, then you can grab things like `parse_state()` and `root()`.